sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala

```diff
@@ -116,13 +116,26 @@ object FileFormatWriter extends Logging {
     val partitionSet = AttributeSet(partitionColumns)
     val dataColumns = allColumns.filterNot(partitionSet.contains)
 
-    val bucketIdExpression = bucketSpec.map { spec =>
+    val bucketPartitioning = bucketSpec.map { spec =>
       val bucketColumns = spec.bucketColumnNames.map(c => dataColumns.find(_.name == c).get)
+      HashPartitioning(bucketColumns, spec.numBuckets)
+    }
+
+    val bucketIdExpression = bucketPartitioning.map { partitioning =>
       // Use `HashPartitioning.partitionIdExpression` as our bucket id expression, so that we can
       // guarantee the data distribution is same between shuffle and bucketed data source, which
       // enables us to only shuffle one side when join a bucketed table and a normal one.
-      HashPartitioning(bucketColumns, spec.numBuckets).partitionIdExpression
+      partitioning.partitionIdExpression
     }
+
+    // If the plan's outputPartitioning is the same as the the bucket spec, then each row will have
+    // a constant bucket id. We possibly can avoid the sort altogether.
+    val bucketSortExpression = plan.outputPartitioning match {
+      case output: HashPartitioning =>
+        bucketPartitioning.filterNot(_.semanticEquals(output)).map(_.partitionIdExpression)
```

Comment:
The output partitioning doesn't guarantee that there's only one value for the partitioning columns in the partition.

For example, assume the output partitioning is [a], it only guarantees that the rows with the same values of [a] will be in the same partition. So a partition can contains all the rows with `a = 1`, `a = 2`..., rather than all rows in the partition have `a = 1`.